policystore/ipset.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policystore
+
+import (
+	"fmt"
+
+	envoyapi "github.com/envoyproxy/data-plane-api/api"
+	syncapi "github.com/projectcalico/app-policy/proto"
+
+	log "github.com/sirupsen/logrus"
+	"strings"
+)
+
+type IPSet interface {
+	// Idempotent add IP address to set.
+	// ip depends on the IPSet type:
+	// IP          - Each member is an IP address in dotted-decimal or IPv6 format.
+	// IP_AND_PORT - Each member is "<IP>,(tcp|udp):<port-number>"
+	AddString(ip string)
+
+	// Idempotent remove IP address from set.
+	// ip depends on the IPSet type:
+	// IP          - Each member is an IP address in dotted-decimal or IPv6 format.
+	// IP_AND_PORT - Each member is "<IP>,(tcp|udp):<port-number>"
+	RemoveString(ip string)
+
+	// Test if the address is contained in the set.
+	ContainsAddress(addr *envoyapi.Address) bool
+}
+
+// We'll use golang's map type under the covers here because it is simple to implement.
+// Later, we may wish to experiment with more memory optimized data structures, like radix trees.
```

Comment:
I think we can use Go interfaces here, `ipMapSet` and `ipPortMapSet` can be its implementations (using map right now), and changing the concrete type to something else in future doesn't affect the code much.

Reply:
I don't understand this comment. I _have_ used a Go interface, `IPSet`

Reply:
Oh sorry my bad, I totally didn't see the interface lines, that's what happens when I do it lat at night